pkg/monitoring/vms/prometheus/prometheus.go

```diff
@@ -0,0 +1,311 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2018 Red Hat, Inc.
+ *
+ */
+
+package prometheus
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"kubevirt.io/kubevirt/pkg/log"
+	"kubevirt.io/kubevirt/pkg/version"
+	cmdclient "kubevirt.io/kubevirt/pkg/virt-handler/cmd-client"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/stats"
+)
+
+const statsMaxAge time.Duration = collectionTimeout + 2*time.Second // "a bit more" than timeout, heuristic again
+
+var (
+	// see https://www.robustperception.io/exposing-the-software-version-to-prometheus
+	versionDesc = prometheus.NewDesc(
+		"kubevirt_info",
+		"Version information",
+		[]string{"goversion", "kubeversion"},
+		nil,
+	)
+
+	storageIopsDesc = prometheus.NewDesc(
+		"kubevirt_vm_storage_iops",
+		"I/O operation performed.",
+		[]string{"domain", "drive", "type"},
+		nil,
+	)
+	// from now on: TODO: validate
+	vcpuUsageDesc = prometheus.NewDesc(
+		"kubevirt_vm_vcpu_time",
+		"Vcpu elapsed time, seconds.",
+		[]string{"domain", "id", "state"},
+		nil,
+	)
+	networkTrafficDesc = prometheus.NewDesc(
+		"kubevirt_vm_network_traffic_bytes",
+		"network traffic, bytes.",
+		[]string{"domain", "interface", "type"},
+		nil,
+	)
+	memoryUsageDesc = prometheus.NewDesc(
+		"kubevirt_vm_memory_amount_bytes",
```

Comment:
I am unsure if `kubevirt_vm_memory_amount_bytes{domain="default_vmi-fedora",type="resident"} 911248` is the best option. Are there other memory types which would then make summing up memory values not meaningful? If yes I would prefer `kubevirt_vm_memory_resident_bytes`, and so on. The case I am thinking about is e.g. reporting reserved and real memory. If I would sum it up, I would get reserved + real.

Reply:
Great that you mentioned that, because this was *exactly* what I was about to bring up next. The current names are my attempt to follow the best practices, but I got no real review about naming/labeling scheme, so I'm very open to suggestions.

Reply:
When I look at your other values:

```
kubevirt_vm_network_traffic_bytes{ type="rx" }
kubevirt_vm_network_traffic_bytes{ type="tx" }
```

This defintely has meaning summing it up gives you the whole traffic of the interface.

And for disks:

```
kubevirt_vm_storage_iops{domain="default_vmi-fedora",drive="vda",type="flush"} 317
kubevirt_vm_storage_iops{domain="default_vmi-fedora",drive="vda",type="read"} 7205
kubevirt_vm_storage_iops{domain="default_vmi-fedora",drive="vda",type="write"} 1329
```

I think it kind of has a meaning. It is not necessarily useful (e.g. read + write is useful, but read + flush probably not) but it still makes sense (in the sense of all disk related operations). That is fine for me. But in the case of memory we might not have that. a `sum()` or an `avg()` on `virtual` and `real` memory for instance is not just not useful it is conceptually wrong.

I don't know the other memory values which might come.

Reply:
Btw. here a list of the metrics which I exposed in the vAdvisor experiment: https://github.com/kubevirt/vAdvisor/blob/master/vadvisor/app/prometheus.py

Reply:
Good point, later today I'll review the vAdvisor metrics and address your comments here.

Reply:
I see at least the following for memory `actual`, `swap_in`, `rss` (resident)

Reply:
Actual and Resident are already readily available. I'll split the memory metric like you suggested and add
- kubevirt_vm_memory_available_bytes
- kubevirt_vm_memory_rss_bytes

I will add more meory stats in a followup PR

Reply:
Here: https://github.com/kubevirt/kubevirt/issues/2023